File: server-node/config.test.js
'use strict';

const assert = require('assert');
const { INSECURE_DEFAULT_SECRET, signingSecret, signingSecretFromEnv } = require('./config');

assert.throws(() => signingSecretFromEnv({}), /FCAPTCHA_SECRET is required/);
assert.throws(() => signingSecretFromEnv({ FCAPTCHA_SECRET: INSECURE_DEFAULT_SECRET }), /FCAPTCHA_SECRET is required/);
assert.strictEqual(signingSecretFromEnv({ FCAPTCHA_SECRET: 'a-real-deployment-secret-0123456789abcdef0123456789abcdef' }), 'a-real-deployment-secret-0123456789abcdef0123456789abcdef');
assert.strictEqual(signingSecretFromEnv({ FCAPTCHA_INSECURE_DEV_MODE: 'true' }), INSECURE_DEFAULT_SECRET);
assert.strictEqual(signingSecret('explicit-library-secret-0123456789abcdef0123456789abcdef', {}), 'explicit-library-secret-0123456789abcdef0123456789abcdef');
assert.throws(() => signingSecret(undefined, {}), /FCAPTCHA_SECRET is required/);
for (const secret of ['x', 'my-secret', 'a'.repeat(64), 'abcd'.repeat(16)]) {
  assert.throws(() => signingSecretFromEnv({ FCAPTCHA_SECRET: secret }), /FCAPTCHA_SECRET is required/);
}

console.log('signing-secret configuration tests passed');

File: server-go/config_test.go
package main

import "testing"

func configEnv(values map[string]string) func(string) string {
	return func(key string) string { return values[key] }
}

func TestSigningSecretFailsClosed(t *testing.T) {
	for _, values := range []map[string]string{
		{},
		{"FCAPTCHA_SECRET": insecureDefaultSecret},
		{"FCAPTCHA_SECRET": "x"},
		{"FCAPTCHA_SECRET": "my-secret"},
		{"FCAPTCHA_SECRET": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"},
	} {
		if _, err := signingSecretFromEnv(configEnv(values)); err == nil {
			t.Fatalf("configuration %v should fail", values)
		}
	}
}

func TestSigningSecretAcceptsConfiguredSecret(t *testing.T) {
	got, err := signingSecretFromEnv(configEnv(map[string]string{"FCAPTCHA_SECRET": "a-real-deployment-secret-0123456789abcdef0123456789abcdef"}))
	if err != nil || got != "a-real-deployment-secret-0123456789abcdef0123456789abcdef" {
		t.Fatalf("got secret=%q err=%v", got, err)
	}
}

func TestSigningSecretExplicitDevelopmentMode(t *testing.T) {
	got, err := signingSecretFromEnv(configEnv(map[string]string{"FCAPTCHA_INSECURE_DEV_MODE": "yes"}))
	if err != nil || got != insecureDefaultSecret {
		t.Fatalf("got secret=%q err=%v", got, err)
	}
}
